lib/rbac/filterer.rb

```diff
@@ -417,17 +402,33 @@ def scope_to_tenant(scope, user, miq_group)
       tenant_id_clause ? scope.where(tenant_id_clause) : scope
     end
 
+    ##
+    # Main scoping method
+    #
     def scope_targets(klass, scope, rbac_filters, user, miq_group)
+      # Results are scoped by tenant if the TenancyMixin is included in the class,
+      # with a few manual exceptions (User, Tenant). Note that the classes in
+      # TENANT_ACCESS_STRATEGY are a consolidated list of them.
       if klass.respond_to?(:scope_by_tenant?) && klass.scope_by_tenant?
         scope = scope_to_tenant(scope, user, miq_group)
       end
 
-      if apply_rbac_to_class?(klass)
-        scope = scope_by_direct_rbac(scope, rbac_filters, user, miq_group)
-      elsif apply_rbac_to_associated_class?(klass)
-        scope = scope_by_indirect_rbac(scope, rbac_filters, user, miq_group)
-      elsif apply_user_group_rbac_to_class?(klass, miq_group)
-        scope = scope_by_user_group_rbac(scope, user, miq_group)
+      if apply_rbac_directly?(klass)
+        filtered_ids = calc_filtered_ids(scope, rbac_filters, user, miq_group)
+        scope = scope_by_ids(scope, filtered_ids)
+      elsif apply_rbac_through_association?(klass)
+        # if subclasses of MetricRollup or Metric, use the associated
+        # model to derive permissions from
+        associated_class = rbac_class(scope)
+        filtered_ids = calc_filtered_ids(associated_class, rbac_filters, user, miq_group)
+
+        scope = scope_by_parent_ids(associated_class, scope, filtered_ids)
+      elsif klass == User && user.try!(:self_service?)
+        # Self service users searching for users only see themselves
+        scope = scope.where(:id => user.id)
+      elsif klass == MiqGroup && miq_group.try!(:self_service?)
+        # Self Service users searching for groups only see their group
+        scope = scope.where(:id => miq_group.id)
```

Comment:
Note that I know this code isn't exactly pretty. More refactoring to follow, where much of this will be way better, but this is MUCH easier to understand contained within `scope_targets` like this than chasing spaghetti methods.